File: sdk/python/src/agent_mesh/identity.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""A mesh identity is an ed25519 key pair. Its peer ID is the one libp2p
derives, so the same key works in the SDK, in sam-node and on the wire."""

from __future__ import annotations

import os

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from . import base58

PUBLIC_KEY_SIZE = 32
SEED_SIZE = 32

# libp2p crypto.proto PublicKey{Type: Ed25519 (1), Data: <32 bytes>}.
_LIBP2P_PUBLIC_KEY_PREFIX = bytes([0x08, 0x01, 0x12, 0x20])
# libp2p crypto.proto PrivateKey{Type: Ed25519 (1), Data: <seed || public>}.
_LIBP2P_PRIVATE_KEY_PREFIX = bytes([0x08, 0x01, 0x12, 0x40])
# multihash: identity function (0x00), digest length 36.
_IDENTITY_MULTIHASH_PREFIX = bytes([0x00, 0x24])


def libp2p_public_key(public_key_raw: bytes) -> bytes:
    """The libp2p protobuf encoding of an ed25519 public key."""
    if len(public_key_raw) != PUBLIC_KEY_SIZE:
        raise ValueError(f"ed25519 public key must be {PUBLIC_KEY_SIZE} bytes, got {len(public_key_raw)}")
    return _LIBP2P_PUBLIC_KEY_PREFIX + public_key_raw


def peer_id_from_public_key(public_key_raw: bytes) -> str:
    """The peer ID libp2p derives from an ed25519 public key (base58btc, "12D3Koo...")."""
    return base58.encode(_IDENTITY_MULTIHASH_PREFIX + libp2p_public_key(public_key_raw))


def verify_ed25519(public_key_raw: bytes, data: bytes, signature: bytes) -> bool:
    """Verifies an ed25519 signature with a raw 32-byte public key."""
    if len(public_key_raw) != PUBLIC_KEY_SIZE:
        return False
    try:
        Ed25519PublicKey.from_public_bytes(public_key_raw).verify(signature, data)
    except (InvalidSignature, ValueError):
        return False
    return True


class Identity:
    """An ed25519 key pair and the libp2p peer ID it maps to."""

    __slots__ = ("_private", "_seed", "public_key_raw", "peer_id")

    def __init__(self, seed: bytes):
        if len(seed) != SEED_SIZE:
            raise ValueError(f"ed25519 seed must be {SEED_SIZE} bytes, got {len(seed)}")
        self._seed = bytes(seed)
        self._private = Ed25519PrivateKey.from_private_bytes(self._seed)
        self.public_key_raw: bytes = self._private.public_key().public_bytes(
            serialization.Encoding.Raw, serialization.PublicFormat.Raw
        )
        self.peer_id: str = peer_id_from_public_key(self.public_key_raw)

    @classmethod
    def generate(cls) -> "Identity":
        return cls(os.urandom(SEED_SIZE))

    @classmethod
    def from_seed(cls, seed: bytes) -> "Identity":
        return cls(seed)

    @classmethod
    def from_libp2p_private_key(cls, data: bytes) -> "Identity":
        """Loads the libp2p protobuf private key encoding, the format sam-node persists."""
        prefix = len(_LIBP2P_PRIVATE_KEY_PREFIX)
        if len(data) != prefix + 64 or not data.startswith(_LIBP2P_PRIVATE_KEY_PREFIX):
            raise ValueError("not a libp2p ed25519 private key")
        identity = cls(data[prefix : prefix + SEED_SIZE])
        if data[prefix + SEED_SIZE :] != identity.public_key_raw:
            raise ValueError("libp2p private key: public half does not match the seed")
        return identity

    def to_libp2p_private_key(self) -> bytes:
        """The libp2p protobuf private key encoding (seed || public key)."""
        return _LIBP2P_PRIVATE_KEY_PREFIX + self._seed + self.public_key_raw

    @property
    def seed(self) -> bytes:
        """The 32-byte ed25519 seed; what a libp2p implementation loads the key from."""
        return self._seed

    @property
    def libp2p_public_key(self) -> bytes:
        """The libp2p protobuf public key encoding, what the control plane stores."""
        return libp2p_public_key(self.public_key_raw)

    def sign(self, data: bytes) -> bytes:
        return self._private.sign(data)

    def verify(self, data: bytes, signature: bytes) -> bool:
        return verify_ed25519(self.public_key_raw, data, signature)

    def __repr__(self) -> str:
        return f"Identity({self.peer_id})"

File: sdk/js/src/identity.ts
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// A mesh identity is an ed25519 key pair. Its peer ID is the one libp2p
// derives, so the same key works in the SDK, in sam-node and on the wire.

import { createPrivateKey, createPublicKey, sign, verify, type KeyObject } from "node:crypto";
import { encodeBase58 } from "./base58.ts";

const PUBLIC_KEY_SIZE = 32;
const SEED_SIZE = 32;

// libp2p crypto.proto PublicKey{Type: Ed25519 (1), Data: <32 bytes>}.
const LIBP2P_PUBLIC_KEY_PREFIX = Uint8Array.of(0x08, 0x01, 0x12, 0x20);
// libp2p crypto.proto PrivateKey{Type: Ed25519 (1), Data: <seed || public>}.
const LIBP2P_PRIVATE_KEY_PREFIX = Uint8Array.of(0x08, 0x01, 0x12, 0x40);
// multihash: identity function (0x00), digest length 36.
const IDENTITY_MULTIHASH_PREFIX = Uint8Array.of(0x00, 0x24);

// PKCS#8 wrapper for a raw ed25519 seed (RFC 8410): what node:crypto imports.
const PKCS8_ED25519_PREFIX = Uint8Array.of(
  0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x04, 0x22, 0x04, 0x20,
);
// SubjectPublicKeyInfo wrapper for a raw ed25519 public key (RFC 8410).
const SPKI_ED25519_PREFIX = Uint8Array.of(0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00);

function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

function startsWith(bytes: Uint8Array, prefix: Uint8Array): boolean {
  return prefix.every((b, i) => bytes[i] === b);
}

function publicKeyObject(publicKeyRaw: Uint8Array): KeyObject {
  return createPublicKey({ key: Buffer.from(concat(SPKI_ED25519_PREFIX, publicKeyRaw)), format: "der", type: "spki" });
}

/** The libp2p protobuf encoding of an ed25519 public key. */
export function libp2pPublicKey(publicKeyRaw: Uint8Array): Uint8Array {
  if (publicKeyRaw.length !== PUBLIC_KEY_SIZE) {
    throw new Error(`ed25519 public key must be ${PUBLIC_KEY_SIZE} bytes, got ${publicKeyRaw.length}`);
  }
  return concat(LIBP2P_PUBLIC_KEY_PREFIX, publicKeyRaw);
}

/** The peer ID libp2p derives from an ed25519 public key (base58btc, "12D3Koo..."). */
export function peerIdFromPublicKey(publicKeyRaw: Uint8Array): string {
  return encodeBase58(concat(IDENTITY_MULTIHASH_PREFIX, libp2pPublicKey(publicKeyRaw)));
}

export class Identity {
  readonly #privateKey: KeyObject;
  readonly #publicKey: KeyObject;
  readonly #seed: Uint8Array;
  /** Raw 32-byte ed25519 public key. */
  readonly publicKeyRaw: Uint8Array;
  readonly peerId: string;

  private constructor(seed: Uint8Array) {
    if (seed.length !== SEED_SIZE) {
      throw new Error(`ed25519 seed must be ${SEED_SIZE} bytes, got ${seed.length}`);
    }
    this.#seed = new Uint8Array(seed);
    this.#privateKey = createPrivateKey({
      key: Buffer.from(concat(PKCS8_ED25519_PREFIX, this.#seed)),
      format: "der",
      type: "pkcs8",
    });
    this.#publicKey = createPublicKey(this.#privateKey);
    const spki = new Uint8Array(this.#publicKey.export({ format: "der", type: "spki" }));
    if (spki.length !== SPKI_ED25519_PREFIX.length + PUBLIC_KEY_SIZE || !startsWith(spki, SPKI_ED25519_PREFIX)) {
      throw new Error("unexpected ed25519 public key encoding");
    }
    this.publicKeyRaw = spki.slice(SPKI_ED25519_PREFIX.length);
    this.peerId = peerIdFromPublicKey(this.publicKeyRaw);
  }

  static generate(): Identity {
    const seed = new Uint8Array(SEED_SIZE);
    crypto.getRandomValues(seed);
    return new Identity(seed);
  }

  static fromSeed(seed: Uint8Array): Identity {
    return new Identity(seed);
  }

  /** Loads the libp2p protobuf private key encoding, the format sam-node persists. */
  static fromLibp2pPrivateKey(bytes: Uint8Array): Identity {
    if (bytes.length !== LIBP2P_PRIVATE_KEY_PREFIX.length + 64 || !startsWith(bytes, LIBP2P_PRIVATE_KEY_PREFIX)) {
      throw new Error("not a libp2p ed25519 private key");
    }
    const seed = bytes.subarray(LIBP2P_PRIVATE_KEY_PREFIX.length, LIBP2P_PRIVATE_KEY_PREFIX.length + SEED_SIZE);
    const pub = bytes.subarray(LIBP2P_PRIVATE_KEY_PREFIX.length + SEED_SIZE);
    const id = new Identity(seed);
    if (!pub.every((b, i) => id.publicKeyRaw[i] === b)) {
      throw new Error("libp2p private key: public half does not match the seed");
    }
    return id;
  }

  /** The libp2p protobuf private key encoding (seed || public key). */
  toLibp2pPrivateKey(): Uint8Array {
    return concat(LIBP2P_PRIVATE_KEY_PREFIX, this.#seed, this.publicKeyRaw);
  }

  /** The libp2p protobuf public key encoding, what the control plane stores. */
  get libp2pPublicKey(): Uint8Array {
    return libp2pPublicKey(this.publicKeyRaw);
  }

  sign(data: Uint8Array): Uint8Array {
    return new Uint8Array(sign(null, data, this.#privateKey));
  }

  verify(data: Uint8Array, signature: Uint8Array): boolean {
    return verify(null, data, this.#publicKey, signature);
  }
}

/** Verifies an ed25519 signature with a raw 32-byte public key. */
export function verifyEd25519(publicKeyRaw: Uint8Array, data: Uint8Array, signature: Uint8Array): boolean {
  if (publicKeyRaw.length !== PUBLIC_KEY_SIZE) {
    return false;
  }
  return verify(null, data, publicKeyObject(publicKeyRaw), signature);
}
